etcdserver/api/membership/cluster.go

```diff
@@ -262,29 +268,41 @@ func (c *RaftCluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
 	}
 	switch cc.Type {
 	case raftpb.ConfChangeAddNode, raftpb.ConfChangeAddLearnerNode:
-		if members[id] != nil {
-			return ErrIDExists
-		}
 		urls := make(map[string]bool)
 		for _, m := range members {
 			for _, u := range m.PeerURLs {
 				urls[u] = true
 			}
 		}
-		m := new(Member)
-		if err := json.Unmarshal(cc.Context, m); err != nil {
+
+		confChangeContext := new(ConfigChangeContext)
+		if err := json.Unmarshal(cc.Context, confChangeContext); err != nil {
 			if c.lg != nil {
-				c.lg.Panic("failed to unmarshal member", zap.Error(err))
+				c.lg.Panic("failed to unmarshal confChangeContext", zap.Error(err))
 			} else {
-				plog.Panicf("unmarshal member should never fail: %v", err)
+				plog.Panicf("unmarshal confChangeContext should never fail: %v", err)
 			}
 		}
-		for _, u := range m.PeerURLs {
-			if urls[u] {
-				return ErrPeerURLexists
+		// A ConfChangeAddNode to a existing learner node promotes it to a voting member.
+		if confChangeContext.IsPromote {
```

Comment:
move the validation to the top of this func?

Reply:
we should do the validation first, then do the context encoding.

Reply:
We need to encode confChangeContext first, which will tells us if the config change is for promoting a existing member, or for adding a new member. The validation logic is different for these two scenarios.

Reply:
ok. sgtm.

Reply:
but then, can you move the decoding thing even before the urls construction?

Reply:
Right, urls construction should be moved to later part of the logic. It is only needed for adding a new member case, not needed for promoting.